Community: Added the MindsDB AI Mind Tool

MinuraPunchihewa opened the pull request:
Greetings LangChain team!

**Description:** 

This PR adds the `AIMindTool` from MindsDB. In essence, this tool allows users to interact with concept of a 'Mind' from MindsDB.

A Mind is defined an AI system with built-in expertise designed to help AI agents accomplish tasks. These plug-and-play systems need little setup and can be customized for advanced use cases.

At the moment, the [AI-Data Mind](https://docs.mdb.ai/docs/minds), which allows SQL databases to be queried in plain English, is supported with more coming very soon.

Given below is an example of how this tool can be used:

```
import os
import getpass

from langchain_community.tools import AIMindTool
from langchain_community.utilities import AIDataMindWrapper


os.environ["MINDS_API_KEY"] = getpass.getpass("Enter your MINDS API key: ")
os.environ["DATABASE_PASSWORD"] = getpass.getpass("Enter your database password: ")

ai_data_mind_tool = AIMindTool(
    api_wrapper=AIDataMindWrapper(
        data_source_configs=[
            {
                'type': 'postgres',
                'description': 'House Sales',
                'connection_args': {
                    'user': 'demo_user',
                    'host': 'samples.mindsdb.com',
                    'port': 5432,
                    'database': 'demo',
                    'schema': 'demo_data'
                }
            }
        ]
    )
)

ai_data_mind_tool.run("How many 2 bedroom houses have been sold?")
```

As more Minds (such as the Finance Mind) are released, more wrappers will be introduced that can be initialized and passed to this same tool.

I wanted to ask if project structure for this tool is OK? I noticed that there aren't many packages in the `utilities` directory, but I thought to add one as it makes sense to break my code into multiple modules. Further, this makes it easier to contribute a new wrapper when a Mind is released.

**Dependencies:**
The `openai` and `minds-sdk` packages.

**Twitter Handle:**
MindsDB

**Docs:**
Docs have been added in the form of a sample notebook!

**Tests:**
Unit tests have been added!

**Linting:**
Done!

MinuraPunchihewa committed:
added the skeleton for the Database Mind Wrapper class

MinuraPunchihewa committed:
added a few simple validations

MinuraPunchihewa committed:
implemented the method for creating a mind

MinuraPunchihewa committed:
implemented the method for querying a mind

MinuraPunchihewa committed:
implemented the method for running the wrapper class

MinuraPunchihewa committed:
added the mindsdb package for tools

MinuraPunchihewa committed:
added the MindsDB Tool implementation

MinuraPunchihewa committed:
imported the MindsDB Tool to main package

MinuraPunchihewa committed:
corrected name of mindsdb package

MinuraPunchihewa committed:
set default values for pvt params in wrapper

MinuraPunchihewa committed:
called func to get secret value of API key in uses

MinuraPunchihewa committed:
renamed mindsdb.py to database_mind_wrapper.py and moved to mindsdb pkg

MinuraPunchihewa committed:
moved PostgresConnection to database_models.py

MinuraPunchihewa committed:
added the model for MySQL

MinuraPunchihewa committed:
added a method to validate data sources against their connection args

MinuraPunchihewa committed:
used the method to validate connection args

MinuraPunchihewa committed:
imported the database mind wrapper to the main pkg

MinuraPunchihewa committed:
fixed imports to make wrapper accessible from main pkg

MinuraPunchihewa committed:
added a base mind wrapper with common mind args

MinuraPunchihewa committed:
added funcs to get supported data sources and validate creds

MinuraPunchihewa committed:
improved database mind wrapper with use of base mind wrapper and utility funcs

MinuraPunchihewa committed:
renamed MindsDBTool to AIMindTool

MinuraPunchihewa committed:
fixed the import of AIMindTool

MinuraPunchihewa committed:
added common validations to base mind wrapper

MinuraPunchihewa committed:
added minds specific validations to database mind wrapper

MinuraPunchihewa committed:
allowed database creds to be passed as env vars

MinuraPunchihewa committed:
updated logic for validating database creds

MinuraPunchihewa committed:
improved func for getting supported data sources

MinuraPunchihewa committed:
added the database models for MariaDB and ClickHouse

MinuraPunchihewa committed:
updated the validation to generate name if not provided

MinuraPunchihewa committed:
added the models for Snowflake and BigQuery

MinuraPunchihewa committed:
implemented a dict() method for other models

MinuraPunchihewa committed:
added the docs for the tool

MinuraPunchihewa committed:
renamed Database Mind to AI Data Mind

MinuraPunchihewa committed:
added a DataSourceConfig model to allow multiple data sources

MinuraPunchihewa committed:
fixed the module lookup for the AIDataMindWrapper

MinuraPunchihewa committed:
fixed the model for Snowflake

MinuraPunchihewa committed:
updated the documentation for AI Data Mind

MinuraPunchihewa committed:
Merge branch 'master' into feature/mindsdb_database_mind_tool

MinuraPunchihewa committed:
removed the use of root_validator in database models

MinuraPunchihewa committed:
removed the use of root_validator in the mind wrapper and data source config

MinuraPunchihewa committed:
moved the import of DatabaseConfig from mindsdb_sdk

MinuraPunchihewa committed:
updated the database models with their own specific env vars

MinuraPunchihewa committed:
updated the documentation with the new use of env vars

baskaryan commented on libs/community/langchain_community/tools/mindsdb/tool.py:
```diff
@@ -0,0 +1,20 @@
+class AIMindTool(BaseTool):
+    name: Text = "ai_mind"
+    description: Text = (
+
```
missing a description

MinuraPunchihewa replied:
Hey @baskaryan,
Thank you for taking the time to review this PR. I have added a description and also resolved the lint issues as best I can. Please let me know if you have any other comments.

baskaryan committed:
Merge branch 'master' into feature/mindsdb_database_mind_tool

MinuraPunchihewa committed:
added a description for the tool

MinuraPunchihewa committed:
fixed lint issues

MinuraPunchihewa committed:
fixed lint issues in docs

MinuraPunchihewa committed:
fixed a circular import

MinuraPunchihewa committed:
fixed grammar of the tool description

MinuraPunchihewa committed:
fixed a broken link in the docs

baskaryan committed:
Merge branch 'master' into feature/mindsdb_database_mind_tool

baskaryan committed:
Update ai_mind.ipynb

MinuraPunchihewa committed:
added type annotations to funcs missing them

MinuraPunchihewa committed:
added an abstract run() func to base mind wrapper

MinuraPunchihewa committed:
added the initial unit tests for AIDataMindWrapper

MinuraPunchihewa committed:
added the unit tests for running AIDataMindWrapper

MinuraPunchihewa committed:
initialized data sources as a config var

MinuraPunchihewa committed:
removed unused code in unit tests for AIDataMindWrapper

MinuraPunchihewa committed:
added a simple unit test for AIMindTool

MinuraPunchihewa committed:
updated the connection args of database models

MinuraPunchihewa committed:
updated the API wrappers to use minds-sdk

MinuraPunchihewa committed:
made a few more updates to support the use of the minds-sdk

MinuraPunchihewa committed:
updated the docs based on the changes to the SDK

MinuraPunchihewa committed:
updated the unit tests based on changes to the SDK

MinuraPunchihewa committed:
fixed lint issues

MinuraPunchihewa committed:
fixed a couple of links in the docs

MinuraPunchihewa commented:
Hey @baskaryan,
I had to make some slight changes because we released a new SDK for our Minds product.
I've also added unit tests and fixed the outstanding lint issues. Is there anything else I need to do in order to get this merged?

efriis commented:
will let bagatur take a look post changes! 

Also, if you're interested in maintaining this integration without us in the loop, we'd love to get an integration package out! Future PRs against langchain would just be {docs updates, as well as registering your package in `libs/packages.yml`, deprecating this community integration in favor of your integration package}

Here's the guide, and if you have questions, feel free to leave them in the comments on those pages so others can see them! https://python.langchain.com/docs/contributing/how_to/integrations/

MinuraPunchihewa commented:
> will let bagatur take a look post changes!
> 
> Also, if you're interested in maintaining this integration without us in the loop, we'd love to get an integration package out! Future PRs against langchain would just be {docs updates, as well as registering your package in `libs/packages.yml`, deprecating this community integration in favor of your integration package}
> 
> Here's the guide, and if you have questions, feel free to leave them in the comments on those pages so others can see them! https://python.langchain.com/docs/contributing/how_to/integrations/

Hey @efriis,
I am happy to work on that. Would it be OK if I took that up as a second step once this integration is merged?

efriis commented:
would strongly prefer going straight to integration package if you're ok with it, because merging would mean we would need to deprecate this one, and it doesn't make much sense to merge it in with the intention of deprecating right away! Then it just becomes dead code that could have issues in the long-run

let us know if that works for you - bagatur is a bit busy so his review might take longer than releasing the separate package, which I'm happy to review the docs for!

MinuraPunchihewa commented:
> would strongly prefer going straight to integration package if you're ok with it, because merging would mean we would need to deprecate this one, and it doesn't make much sense to merge it in with the intention of deprecating right away! Then it just becomes dead code that could have issues in the long-run
> 
> let us know if that works for you - bagatur is a bit busy so his review might take longer than releasing the separate package, which I'm happy to review the docs for!

OK, understood. I will create the package and let you know.

ccurme commented:
Closing in light of discussion above.
